flutterproject/internal/sdk/asdf.go

```diff
@@ -21,43 +27,57 @@ func NewASDFVersionReader(fileOpener FileOpener) ASDFVersionReader {
 	}
 }
 
-func (r ASDFVersionReader) ReadSDKVersions(projectRootDir string) (*semver.Version, error) {
+func (r ASDFVersionReader) ReadSDKVersions(projectRootDir string) (*semver.Version, string, error) {
 	asdfConfigPth := filepath.Join(projectRootDir, asdfConfigRelPath)
 	f, err := r.fileOpener.OpenReaderIfExists(asdfConfigPth)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 
 	if f == nil {
-		return nil, nil
+		return nil, "", nil
 	}
 
-	versionStr, err := parseASDFFlutterVersion(f)
+	versionStr, channel, err := parseASDFFlutterVersion(f)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 	if versionStr == "" {
-		return nil, nil
+		return nil, "", nil
+	}
+
+	version, err := semver.NewVersion(versionStr)
+	if err != nil {
+		return nil, "", err
 	}
 
-	return semver.NewVersion(versionStr)
+	return version, channel, nil
 }
 
-func parseASDFFlutterVersion(asdfConfigReader io.Reader) (string, error) {
+func parseASDFFlutterVersion(asdfConfigReader io.Reader) (string, string, error) {
+	versionStr := ""
+	channel := ""
+
 	scanner := bufio.NewScanner(asdfConfigReader)
 	scanner.Split(bufio.ScanLines)
-	versionStr := ""
 	for scanner.Scan() {
 		line := scanner.Text()
 		line = strings.TrimSpace(line)
 		if strings.HasPrefix(line, "flutter ") {
 			versionStr = strings.TrimPrefix(line, "flutter ")
+			for _, c := range []string{stableChannel, betaChannel, masterChannel} {
```

Comment:
The pre-release version (stable,alpha...) can be valid semver, do we need to parse it seperately? https://semver.org/#spec-item-9

Reply:
I'm not sure if I got the question correctly.

It seems Flutter SDK versions on all the different release channels (stable, beta, dev) are valid semver versions.

Usually on beta and dev channels the version has the `.pre` suffix, like: `2.11.0-0.1.pre` and the semver lib we use can parse it properly, here is a test case, where a similar version gets parsed: https://github.com/bitrise-io/go-flutter/pull/5/files#diff-4271b57099af43ba18617a832964e7917905e6eb7f0af9b2aa5c97758086aafcR35.
Another example is the `+hotifix` version suffix, like: `1.9.1+hotfix.4`, this is parsed properly as well.

This version `1.9.1+hotfix.4` is available both on the stable and beta channel, if you want to specify from which channel you want to use it, in your asdf config you write `1.9.1+hotfix.4-stable` or `1.9.1+hotfix.4-beta`. This channel suffix is what needs to be cut off, before parsing the version.